### PR TITLE
Add an option to report deprecations

### PR DESCRIPTION
**Pull request by jderusse**
Converts deprecation triggered by vendors (or the application it self) to a dedicated `DeprecationException`

It helps a lots to spot deprecation on a legacy application with few code coverage

* [x] Test (waiting for test refactoring to be merged)
* [x] Update changelog

### CODE REVIEW

File: Logging/DeprecationHandler.php [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+/*
+ * This file is part of Ekino New Relic bundle.
+ *
+ * (c) Ekino - Thomas Rabaix <thomas.rabaix@ekino.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Ekino\Bundle\NewRelicBundle\Logging;
```
Comment (jderusse) [inline Logging/DeprecationHandler.php]: I'm not sure about the namespace. Any suggestion?

**Comment by Nyholm** [comment]
Ooh, cool. I like this. I'll do a review and test it later. (just rushing away).

File: Logging/DeprecationHandler.php (hunk 1 above)
Comment (Nyholm) [reply Logging/DeprecationHandler.php]: Maybe we should consider this as a "Listener" by moving it to the "Listener" namespace?

File: Resources/config/services.xml [hunk 2]
```diff
@@ -80,5 +80,9 @@
             <argument type="service" id="ekino.new_relic" />
             <tag name="console.command" command="newrelic:notify-deployment" />
         </service>
+
+        <service id="ekino.new_relic.deprecation_handler" class="Ekino\Bundle\NewRelicBundle\Logging\DeprecationHandler">
```
Comment (Nyholm) [inline Resources/config/services.xml]: You need to make it public if it going to work in Symfony 4.

File: Logging/DeprecationHandler.php [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    }
+
+    public function register(): void
+    {
+        if ($this->isRegistered) {
+            return;
+        }
+        $this->isRegistered = true;
+
+        $prevErrorHandler = set_error_handler(function ($type, $msg, $file, $line, $context = array()) use (&$prevErrorHandler) {
```
Comment (Nyholm) [inline Logging/DeprecationHandler.php]: This looks super cool. But does it really work with `$prevErrorHandler = .... use (&$prevErrorHandler)...` like that?

Comment (jderusse) [reply Logging/DeprecationHandler.php]: Let me right some tests to prove it :)

File: Tests/Listener/DeprecationListenerTest.php [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+        $interactor = $this->getMockBuilder(NewRelicInteractorInterface::class)->getMock();
+        $interactor->expects($this->never())->method('noticeException');
+
+        $listener = new DeprecationListener($interactor);
+
+        set_error_handler(function () { return false; });
+        try {
+            $listener->register();
+            $listener->unregister();
+            @trigger_error('This is a notice', E_USER_NOTICE);
```
Comment (jderusse) [inline Tests/Listener/DeprecationListenerTest.php]: This should be a deprecation

Comment (Nyholm) [reply Tests/Listener/DeprecationListenerTest.php]: What do you mean?

Comment (jderusse) [reply Tests/Listener/DeprecationListenerTest.php]: Don't worry, I also comment my code (as a reminder).

The purpose of this test is to assert that when calling `unregister`, the deprecation is not pushed to new relic. But given, this line is triggering a `E_USER_NOTICE` the event won't, in any way, be pushed to newrelic

**Comment by jderusse** [comment]
Ready for review. If someone may test it with reall metric and give feedback about the format of reported deprecation :)

**Comment by Nyholm** [comment]
Im happy with this. Merge if you like it. 

I'll make sure to test it in production before we release 2.0

**Pull request merged**